stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -2,30 +2,315 @@
 
 package com.facebook.stetho.inspector.protocol.module;
 
+import android.util.SparseArray;
+
+import com.facebook.stetho.common.Util;
+import com.facebook.stetho.inspector.helper.ChromePeerManager;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcPeer;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcResult;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsDomain;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsMethod;
+import com.facebook.stetho.json.ObjectMapper;
 import com.facebook.stetho.json.annotation.JsonProperty;
 
 import org.json.JSONObject;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
 public class DOM implements ChromeDevtoolsDomain {
-  public DOM() {
+  private final Provider mDOMProvider;
+  private final DOMPeerManager mPeerManager;
+  private final ObjectMapper mObjectMapper;
+
+  private final Object mNodeMapSync = new Object();
+
+  @GuardedBy("mNodeMapSync")
+  private int mNextNodeId = 1;
+
+  @GuardedBy("mNodeMapSync")
+  private final Map<Object, Integer> mElementToNodeIdMap = new HashMap<>();
+
+  @GuardedBy("mNodeMapSync")
+  private final SparseArray<Object> mNodeIdToElementMap = new SparseArray<>();
+
+  public DOM(Provider provider) {
```

Comment:
Oh, I think I misspoke previously on my advice about accepting only `Context` here.  This probably is best expressed as this generic interface :)